devel/addon/haproxy/install.sh

```diff
@@ -0,0 +1,73 @@
+#!/usr/bin/env bash
+
+# Copyright 2021 The cert-manager Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Install HAProxy as a gateway-API e2e test.
+
+set -o nounset
+set -o errexit
+set -o pipefail
+
+SCRIPT_ROOT=$(dirname "${BASH_SOURCE}")
+source "${SCRIPT_ROOT}/../../lib/lib.sh"
+
+check_tool helm
+check_tool kubectl
+
+helm repo add haproxy-ingress https://haproxy-ingress.github.io/charts
+helm repo update
+
+export NAMESPACE="haproxy-ingress"
+export VERSION="0.13.0-beta.2"
```

Comment:
I was not able to use haproxy-ingress due to a bug that I reported this morning (https://github.com/jcmoraisjr/haproxy-ingress/issues/830). Would it be worth it switching to traefik in the meantime?

The script I use to install the correct things as per the gateway API traefik documentation at https://doc.traefik.io/traefik/routing/providers/kubernetes-gateway/:

```sh
helm upgrade --install traefik traefik/traefik --namespace traefik --create-namespace \
  --set additionalArguments='{--providers.kubernetesingress,--providers.kubernetesingress.ingressendpoint.publishedservice=traefik/traefik,--experimental.kubernetesgateway=true,--providers.kubernetesgateway=true}' \
  --version 9.20.1

kubectl apply -f- <<EOF
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: gateway-role
rules:
  - apiGroups:
      - ""
    resources:
      - services
      - endpoints
      - secrets
    verbs:
      - get
      - list
      - watch
  - apiGroups:
      - networking.x-k8s.io
    resources:
      - gatewayclasses
      - gateways
      - httproutes
    verbs:
      - get
      - list
      - watch
  - apiGroups:
      - networking.x-k8s.io
    resources:
      - gatewayclasses/status
      - gateways/status
      - httproutes/status
    verbs:
      - update
---
kind: ClusterRoleBinding
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: gateway-controller
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: gateway-role
subjects:
  - kind: ServiceAccount
    name: traefik
    namespace: traefik
EOF
```

The gateway class looks like this:

```yaml
apiVersion: networking.x-k8s.io/v1alpha1
kind: GatewayClass
metadata:
  name: traefik
spec:
  controller: traefik.io/gateway-controller # ✅
```

Reply:
@jakexks We can't currently use Traefik for our end-to-end tests due to the fact that Traefik only works with `allow: Same`, and `allow: All` is ignored (https://github.com/traefik/traefik/issues/8246).

This issue should be fixed very shortly, should I still keep things around?

Currently, you can run the e2e tests with Traefik using:
 
```sh
./devel/addon/traefik/install.sh
go test ./test/e2e/ -ginkgo.focus 'Certificates with issuer type ACME HTTP01 Issuer \(Gateway\)' \
  --test.v \
  --repo-root $PWD/devel/lib/../.. \
  --report-dir $PWD/devel/lib/../../_artifacts \
  --gateway-domain=traefik.http01.example.com \
  --gateway-httproute-labels=acme=solver-traefik
```

Reply:
> ./devel/addon/traefik/install.sh

FWIW when I run this script, the bit where we read helm values from stdin fails for me locally with `Error: read /dev/stdin: permission denied` .

Reply:
I just realized that helm [supports](https://github.com/helm/helm/issues/7002) using `--values -`, but not sure whether that would fix the issue.

@irbekrm Should I rather use `--set` for all the options? Or store `values.yaml` into a file?

@jakexks Should we revert the two commits 345ace6 and f96f910 since we aren't going to be using Traefik in the end-to-end tests?

Reply:
> I just realized that helm supports using --values -, but not sure whether that would fix the issue.

It didn't fix for me, not entirely sure if something is off for me locally though. To me a values file would make sense, we already have some here https://github.com/jetstack/cert-manager/tree/master/test/fixtures, but up to you- I initially didn't realize that we don't anyway use traefik for the e2e tests?